docs/HowToFAQ/Errors-Compiler.md

```diff
@@ -1,104 +1,529 @@
+Hence this 'assign any legal value' (`... := *`) statement is not permitted with `--enforce-determinism`.
+
+There are a few different forms of this kind of assignment:
+- `x := *;`
+- `x, y, z := 3, *, 42;`
+- `forall i | 0 <= i < a.Length { a[i] := *; }`
+
+## **Error: assign-such-that statement forbidden by the --enforce-determinism option**
+
+<!-- %check-run %options --enforce-determinism -->
+```dafny
+method m() {
+  var x: int;
+  x :| x < 5;
+}
+```
 
-## nondeterministic loop forbidden by the --enforce-determinism option
+The `--enforce-determinism` option requires a target program to be deterministic and predictable:
+there may be no program statements that have an arbitrary, even if deterministic, result.
+Hence this 'assign any value that satisfies the predicate' (`:|`) statement is not permitted with `--enforce-determinism`,
+even if there is only one such possible value.
 
-## case-based loop forbidden by the --enforce-determinism option
 
-## nondeterministic assignment forbidden by the --enforce-determinism option
+## **Error: this assign-such-that statement is too advanced for the current compiler; Dafny's heuristics cannot find any bound for variable '_name_'**
 
--- this is the third occurrence
+TODO
 
-## compiler currently does not support assignments to more-than-6-dimensional arrays in forall statements
+## **Error: nondeterministic if statement forbidden by the --enforce-determinism option**
 
-## modify statement without a body forbidden by the --enforce-determinism option
+<!-- %check-run %options --enforce-determinism -->
+```dafny
+method m() {
+var y: int;
+  if * { y:= 0; } else { y:= 1; }
+}
+```
+
+The `--enforce-determinism` option requires a target program to be deterministic and predictable:
+there may be no program statements that have an arbitrary, even if deterministic, result.
+Hence this 'non-deterministic if' (`if *`) statement is not permitted with `--enforce-determinism`.
+
+## **Error: binding if statement forbidden by the --enforce-determinism option**
+
+<!-- TODO: a binding guard is a ghost statement so how could this situation ever happen? -->
+
+<!-- %no-check %options --enforce-determinism -->
+```dafny
+method m(k: int) {
+  var i := k;
+  var y: int;
+  if i :| i < 5 {  } else { y := 1; }
+}
+```
+
+The `--enforce-determinism` option requires a target program to be deterministic and predictable:
+there may be no program statements that have an arbitrary, even if deterministic, result.
+The [`if` with a binding guard](../DafnyRef/DafnyRef#sec-binding-guards) (`if ... :| `) checks that the 
+given predicate is satisfiable by some value. If not, then the 'else' branch is executed;
+but if so, the 'then' branch is executed with an arbitrary value that satisifies the predicate.
+Because of this arbitrary selection, the if-with-binding-guard is not permitted with `--enforce-determinism`,
+even if there is exactly one value that satisfies the predicate.
```

Comment:
Again, it could be worth having a sentence about why.

Reply:
Same explanation